internal/defence/limiter.go

```diff
@@ -0,0 +1,146 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package defence
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"golang.org/x/time/rate"
+
+	"storj.io/storj/internal/sync2"
+)
+
+// Attacker stores information about attacker entity
+type Attacker struct {
+	limiter *rate.Limiter
+	Expire  time.Time
+	// TODO: could be interface{}
+	key string
+
+	IsBanned bool
+}
+
+func (attacker *Attacker) attack(attemptTime time.Time, banDuration time.Duration) bool {
+	// is not cleared by Clear method of Limiter
+	if attacker.IsBanned && attemptTime.After(attacker.Expire) {
+		attacker.IsBanned = false
+	}
+
+	if attacker.IsBanned {
+		return true
+	}
+
+	isBanned := !attacker.limiter.AllowN(attemptTime, 1)
+
+	if isBanned {
+		attacker.IsBanned = true
+		attacker.Expire = attemptTime.Add(banDuration)
+	}
+
+	return isBanned
+}
+
+// Limiter is used to store and manage list of banned entities
+type Limiter struct {
+	attackers map[string]*Attacker
+
+	// Attempts defines how many times attacker could perform an operation
+	Attempts int
+	// AttemptsPeriod defines period in which attempts will count. For example, 5 attempts per minute.
+	AttemptsPeriod time.Duration
+	BanDuration    time.Duration
+
+	mu   sync.Mutex
+	loop *sync2.Cycle
+}
+
+// NewLimiter is a constructor for Limiter
+func NewLimiter(attempts int, attemptsPeriod, banDuration, clearPeriod time.Duration) *Limiter {
+	return &Limiter{
+		attackers:      map[string]*Attacker{},
+		Attempts:       attempts,
+		AttemptsPeriod: attemptsPeriod,
+		BanDuration:    banDuration,
+		loop:           sync2.NewCycle(clearPeriod),
+	}
+}
+
+// Attack is use to add new fail attack
+func (limiter *Limiter) Attack(key string) bool {
+	limiter.mu.Lock()
+
+	defer limiter.mu.Unlock()
+	now := time.Now()
+
+	// Try to retrieve the
+	if _, found := limiter.attackers[key]; !found {
+		limiter.attackers[key] = &Attacker{
+			key:     key,
+			limiter: rate.NewLimiter(rate.Every(limiter.BanDuration), limiter.Attempts),
+			Expire:  now.Add(limiter.BanDuration),
+		}
+	}
+
+	limit := limiter.attackers[key]
+
+	return limit.attack(now, limiter.BanDuration)
+}
+
+// Banned returns the list of banned attackers
+func (limiter *Limiter) Banned() []*Attacker {
+	limiter.mu.Lock()
+
+	defer limiter.mu.Unlock()
+
+	var attackers []*Attacker
+
+	for _, attacker := range limiter.attackers {
+		if attacker.IsBanned {
+			attackers = append(attackers, attacker)
+		}
+	}
+
+	return attackers
+}
+
+// Find can be used to find an attacker by specified key
+func (limiter *Limiter) Find(key string) (Attacker, bool) {
```

Comment:
Returning the attacker from here sounds dangerous due to potential sharing.

Reply:
it will return struct with 2 public fields - IsBanned and Expired. Moreover i returns attacker by value, not by ref

Reply:
The type contains pointers, hence you are sharing them.

Reply:
ok, and what if i will return new attacker without this pointers?
f.e. 
```
return Attacker {
    Expire:      a.Expire,
    IsBanned: a.IsBanned,
}, true
```

Reply:
Why do you need to return it at all?

Reply:
because we could have not 'rate limited' attacker, who has a few attempt before ban

Reply:
I would expect the whole API be something like:

```
func (limiter *Limiter) Limit(ctx context.Context, now time.Time, key string) bool {}

func Example() {
    limiter := NewLimiter(...)
    if limiter.Limit(ctx, "hello") {
        return // respond with error
    }
}
```

Reply:
And if really need to then `func (limiter *Limiter) CleanExpired(ctx context.Context, now time.Time){}`